polymorphic_value.h

```diff
@@ -136,7 +136,7 @@ template<typename T, polymorphic_value_options Options = polymorphic_value_optio
     operator bool() const { return get() != nullptr; }
 
     // Access the stored object. This is the unique_ptr API to allow for drop in replacement.
-    T* get() { return m_handler.get(m_data); }
+    T *get() { return getHandlerBase().get(m_data); }
```

Comment:
I'm seeing the current failure under GCC: 
```bash
PolymorphicValue/test_polymorphic_value.cpp:49: int main(): Assertion `sv' failed.
```
 
Without adding a wrapper function to ensure the call to the handler virtual methods is made via reference then under GCC this appears to directly call the `handle_base::get' which returns a `nullptr`.

Reply:
Visual Studio appears to do as intended here:

```
--- PolymorphicValue\polymorphic_value.h -----

    // Access the stored object. This is the unique_ptr API to allow for drop in replacement.
    T* get() { return m_handler.get(m_data); }
00007FF73A956D00  mov         qword ptr [rsp+8],rcx  
00007FF73A956D05  push        rdi  
00007FF73A956D06  sub         rsp,30h  
00007FF73A956D0A  mov         rax,qword ptr [this]  
00007FF73A956D0F  add         rax,40h  
00007FF73A956D13  mov         qword ptr [rsp+20h],rax  
00007FF73A956D18  mov         rax,qword ptr [this]           // Load this pointer to RAX
00007FF73A956D1D  mov         rax,qword ptr [rax+40h]   // Add 64 bytes so that it points at the vtable of handler
00007FF73A956D21  mov         rax,qword ptr [rax+10h]   // Add 16 bytes to access 3 entry in the vtable
00007FF73A956D25  mov         qword ptr [rsp+28h],rax   // Store it at a 40-bytes offset from the stack pointer
00007FF73A956D2A  mov         rax,qword ptr [this]  
00007FF73A956D2F  mov         rdx,rax  
00007FF73A956D32  mov         rcx,qword ptr [rsp+20h] // Load this handler pointer to the EXC register
00007FF73A956D37  call        qword ptr [rsp+28h]  // Now call the saved function (with the *this pointer set)
00007FF73A956D3B  add         rsp,30h  
00007FF73A956D3F  pop         rdi  
00007FF73A956D40  ret  
```

I suspect Clang and GCC are correct here, but would need to look at the relevant standard wording to ensure what is the correct behaviour.

Reply:
Interesting. I was wondering if I would need to call std::launder() _everywhere_ as I do update the vtable pointer with placement new. I have not found a standards passage that defines for how long the compiler is allowed to cache the old value of the vtable pointer in optimized builds where both an assignment that does the placement new of the handler and subsequent access may be inlined into linear code, which is where you'd normally have to use std::launder.

But this seems to be another issue, that the compiler devirtualizes the call totally, which now that I look back at this code seems reasonable for it to be able to do, m_handler is a by value member of the base class type after all.

Maybe the only non-UB thing to do here is to implement virtual methods yourself with a function pointer table. Did you check if the getHandlerBase() method gets inlined? Could it be that _if_ it gets inlined the error could recur? If so it'd have to be reimplemented with a function pointer table I guess. But my hunch is that the fact that getHandlerBase() returns by reference will guarantee that the call is not devirtualized.

The correct way to do this is probably `std::launder(&m_handler)->get(m_data);` but of course this can be wrapped in a getHandler helper method. btw the function should be called getHandler, not getHandlerBase.

There are probably more places where this should be fixed.

Please don't change the * placement convention in another programmer's repo.

Reply:
Looking at the vtable in GDB it appears to be correct, the issue is just that by using the dot operator on the value directly the compiler does not have to do virtual dispatch.  Yes, avoiding devirtualisation was the intention behind adding the wrapper method.

Hmmm... I missed the * placement, looks like clang-format on my machine defaults to that, certainly not intentional

Reply:
np